pyquil/api/_qam.py

```diff
@@ -37,11 +42,63 @@ class QAMExecutionResult:
     executable: QuantumExecutable
     """The executable corresponding to this result."""
 
-    readout_data: Mapping[str, Optional[np.ndarray]] = field(default_factory=dict)
-    """Readout data returned from the QAM, keyed on the name of the readout register or post-processing node."""
+    data: ExecutionData
+    """
+    The ``ExecutionData`` returned from the job. Consider using
+    ``QAMExecutionResult#register_map`` or ``QAMExecutionResult#raw_readout_data``
+    to get at the data in a more convenient format.
+    """
+
+    @property
+    def raw_readout_data(self) -> Union[RawQVMReadoutData, RawQPUReadoutData]:
+        """
+        Get the raw result data. This will be a flattened dictionary derived
+        from :class:`qcs_sdk.qvm.QVMResultData` or :class:`qcs_sdk.qpu.QPUResultData` depending on where the
+        job was run. See the respective
+
+        This property should be used when running programs that use features like
+        mid-circuit measurement and dynamic control flow on a QPU, since they can
+        produce irregular result shapes that don't necessarily fit in a
+        rectangular matrix. If the program was run on a QVM, or doesn't use those
+        features, consider using the ``register_map`` property instead.
+        """
+        return self.data.result_data.to_raw_readout_data()
 
-    execution_duration_microseconds: Optional[int] = field(default=None)
-    """Duration job held exclusive hardware access. Defaults to ``None`` when information is not available."""
+    @property
+    def register_map(self) -> Dict[str, Optional[np.ndarray]]:
```

Comment:
Just a question, we could go either way - but in line with @Shadow53 's question below about caching, since these are new methods anyway, should we eschew the `@property` and use `get_register_map` etc to make it clear that work is taking place?

Reply:
I like that idea, updated this and the `raw_readout_data` property.